File: apps/desktop/src/lib/agent-control.ts
import { invoke as tauriInvoke } from '@tauri-apps/api/core';
import type { AgentControllerStateV1 } from './agent-activity';

const CURRENT_PROTOCOL_VERSION = 1 as const;
const STABLE_ID_PATTERN = /^[0-9a-f]{64}$/u;
const POSITIVE_DECIMAL_PATTERN = /^[1-9][0-9]*$/u;
const MAX_U32 = 4_294_967_295;
const MAX_U64 = 18_446_744_073_709_551_615n;

type InvokeCommand = (command: string, arguments_: Record<string, unknown>) => Promise<unknown>;

export type AgentTaskControlActionV1 = 'cancel' | 'replan' | 'resume';
export type AgentTaskControlOutcomeV1 = 'cancelled' | 'replanRequired' | 'resumed';

export interface AgentTaskRecoveryV1 {
  canResume: boolean;
  interruptedToolAttempts: number;
  ledgerRevision: number;
  ledgerStoreVersion: string;
  mutationReconciliationRequired: boolean;
  mutationReplanRequired: boolean;
  publishedSnapshotId: string;
  runSnapshotId: string;
  snapshotChanged: boolean;
  staleEvidenceCount: number;
  state: AgentControllerStateV1;
}

export type AgentTaskRecoveryResultV1 =
  | { status: 'noProject' }
  | { status: 'taskNotFound' }
  | { status: 'ledgerUnavailable' }
  | { currentRevision: number; ledgerRevision: number; status: 'goalRevisionMismatch' }
  | { status: 'activityChanged' }
  | { status: 'runUnavailable' }
  | { state: AgentControllerStateV1; status: 'runNotControllable' }
  | { recovery: AgentTaskRecoveryV1; status: 'available' };

export interface AgentTaskRecoveryResponseV1 {
  protocolVersion: typeof CURRENT_PROTOCOL_VERSION;
  result: AgentTaskRecoveryResultV1;
}

export type AgentTaskControlResultV1 =
  | { status: 'noProject' }
  | { status: 'taskNotFound' }
  | { status: 'ledgerUnavailable' }
  | { currentRevision: number; ledgerRevision: number; status: 'goalRevisionMismatch' }
  | { status: 'activityChanged' }
  | { status: 'runUnavailable' }
  | { state: AgentControllerStateV1; status: 'runNotControllable' }
  | { status: 'mutationReconciliationRequired' }
  | { status: 'resumeRequiresReplan' }
  | {
      interruptedToolAttempts: number;
      ledgerStoreVersion: string;
      outcome: AgentTaskControlOutcomeV1;
      reopenedStepCount: number;
      state: AgentControllerStateV1;
      status: 'applied';
    };

export interface AgentTaskControlResponseV1 {
  protocolVersion: typeof CURRENT_PROTOCOL_VERSION;
  result: AgentTaskControlResultV1;
}

const invokeThroughTauri: InvokeCommand = (command, arguments_) =>
  tauriInvoke<unknown>(command, arguments_);

export async function queryAgentTaskRecovery(
  taskId: string,
  invokeCommand: InvokeCommand = invokeThroughTauri,
): Promise<AgentTaskRecoveryResponseV1> {
  if (!isStableId(taskId)) throw new Error('Agent task recovery identity does not match V1.');
  const request = { protocolVersion: CURRENT_PROTOCOL_VERSION, taskId };
  return parseAgentTaskRecoveryResponseV1(
    await invokeCommand('query_agent_task_recovery', { request }),
  );
}

export async function controlAgentTaskRun(
  taskId: string,
  expectedLedgerRevision: number,
  expectedLedgerStoreVersion: string,
  action: AgentTaskControlActionV1,
  invokeCommand: InvokeCommand = invokeThroughTauri,
): Promise<AgentTaskControlResponseV1> {
  if (
    !isStableId(taskId) ||
    !isPositiveU32(expectedLedgerRevision) ||
    !isPositiveDecimal(expectedLedgerStoreVersion) ||
    !isControlAction(action)
  ) {
    throw new Error('Agent task control request does not match V1.');
  }
  const request = {
    action,
    expectedLedgerRevision,
    expectedLedgerStoreVersion,
    protocolVersion: CURRENT_PROTOCOL_VERSION,
    taskId,
  };
  return parseAgentTaskControlResponseV1(
    await invokeCommand('control_agent_task_run', { request }),
  );
}

export function parseAgentTaskRecoveryResponseV1(payload: unknown): AgentTaskRecoveryResponseV1 {
  const result = parseEnvelope(payload);
  return { protocolVersion: CURRENT_PROTOCOL_VERSION, result: parseRecoveryResult(result) };
}

export function parseAgentTaskControlResponseV1(payload: unknown): AgentTaskControlResponseV1 {
  const result = parseEnvelope(payload);
  return { protocolVersion: CURRENT_PROTOCOL_VERSION, result: parseControlResult(result) };
}

function parseEnvelope(payload: unknown): unknown {
  if (
    !isRecord(payload) ||
    !hasExactKeys(payload, ['protocolVersion', 'result']) ||
    payload.protocolVersion !== CURRENT_PROTOCOL_VERSION
  ) {
    throw new Error('Agent task control response does not match V1.');
  }
  return payload.result;
}

function parseRecoveryResult(value: unknown): AgentTaskRecoveryResultV1 {
  const common = parseCommonResult(value);
  if (common !== null) return common;
  if (
    isRecord(value) &&
    value.status === 'available' &&
    hasExactKeys(value, ['recovery', 'status'])
  ) {
    return { recovery: parseRecovery(value.recovery), status: 'available' };
  }
  throw new Error('Agent task recovery result uses an unsupported state.');
}

function parseControlResult(value: unknown): AgentTaskControlResultV1 {
  const common = parseCommonResult(value);
  if (common !== null) return common;
  if (
    isRecord(value) &&
    (value.status === 'mutationReconciliationRequired' ||
      value.status === 'resumeRequiresReplan') &&
    hasExactKeys(value, ['status'])
  ) {
    return { status: value.status };
  }
  if (
    isRecord(value) &&
    value.status === 'applied' &&
    hasExactKeys(value, [
      'interruptedToolAttempts',
      'ledgerStoreVersion',
      'outcome',
      'reopenedStepCount',
      'state',
      'status',
    ]) &&
    isU32(value.interruptedToolAttempts) &&
    isPositiveDecimal(value.ledgerStoreVersion) &&
    isControlOutcome(value.outcome) &&
    isU32(value.reopenedStepCount) &&
    isControllerState(value.state) &&
    ((value.outcome === 'cancelled' && value.state === 'cancelled') ||
      (value.outcome !== 'cancelled' && !isTerminalState(value.state)))
  ) {
    return value as AgentTaskControlResultV1;
  }
  throw new Error('Agent task control result uses an unsupported state.');
}

function parseCommonResult(
  value: unknown,
): Exclude<AgentTaskRecoveryResultV1, { status: 'available' }> | null {
  if (!isRecord(value) || typeof value.status !== 'string') return null;
  if (
    [
      'noProject',
      'taskNotFound',
      'ledgerUnavailable',
      'activityChanged',
      'runUnavailable',
    ].includes(value.status) &&
    hasExactKeys(value, ['status'])
  ) {
    return { status: value.status } as Exclude<AgentTaskRecoveryResultV1, { status: 'available' }>;
  }
  if (
    value.status === 'goalRevisionMismatch' &&
    hasExactKeys(value, ['currentRevision', 'ledgerRevision', 'status']) &&
    isPositiveU32(value.currentRevision) &&
    isPositiveU32(value.ledgerRevision) &&
    value.currentRevision !== value.ledgerRevision
  ) {
    return {
      currentRevision: value.currentRevision,
      ledgerRevision: value.ledgerRevision,
      status: 'goalRevisionMismatch',
    };
  }
  if (
    value.status === 'runNotControllable' &&
    hasExactKeys(value, ['state', 'status']) &&
    isControllerState(value.state)
  ) {
    return { state: value.state, status: 'runNotControllable' };
  }
  return null;
}

function parseRecovery(value: unknown): AgentTaskRecoveryV1 {
  if (
    !isRecord(value) ||
    !hasExactKeys(value, [
      'canResume',
      'interruptedToolAttempts',
      'ledgerRevision',
      'ledgerStoreVersion',
      'mutationReconciliationRequired',
      'mutationReplanRequired',
      'publishedSnapshotId',
      'runSnapshotId',
      'snapshotChanged',
      'staleEvidenceCount',
      'state',
    ]) ||
    typeof value.canResume !== 'boolean' ||
    !isU32(value.interruptedToolAttempts) ||
    !isPositiveU32(value.ledgerRevision) ||
    !isPositiveDecimal(value.ledgerStoreVersion) ||
    typeof value.mutationReconciliationRequired !== 'boolean' ||
    typeof value.mutationReplanRequired !== 'boolean' ||
    !isStableId(value.publishedSnapshotId) ||
    !isStableId(value.runSnapshotId) ||
    typeof value.snapshotChanged !== 'boolean' ||
    value.snapshotChanged !== (value.publishedSnapshotId !== value.runSnapshotId) ||
    !isU32(value.staleEvidenceCount) ||
    !isControllerState(value.state) ||
    isTerminalState(value.state) ||
    value.canResume !==
      (value.staleEvidenceCount === 0 &&
        !value.mutationReconciliationRequired &&
        !value.mutationReplanRequired) ||
    (value.mutationReconciliationRequired && value.mutationReplanRequired)
  ) {
    throw new Error('Agent task recovery projection is invalid.');
  }
  return value as unknown as AgentTaskRecoveryV1;
}

function isStableId(value: unknown): value is string {
  return typeof value === 'string' && STABLE_ID_PATTERN.test(value);
}

function isPositiveDecimal(value: unknown): value is string {
  return (
    typeof value === 'string' && POSITIVE_DECIMAL_PATTERN.test(value) && BigInt(value) <= MAX_U64
  );
}

function isPositiveU32(value: unknown): value is number {
  return isU32(value) && value > 0;
}

function isU32(value: unknown): value is number {
  return typeof value === 'number' && Number.isInteger(value) && value >= 0 && value <= MAX_U32;
}

function isControllerState(value: unknown): value is AgentControllerStateV1 {
  return [
    'intake',
    'localize',
    'plan',
    'execute',
    'verify',
    'replan',
    'awaitApproval',
    'done',
    'failed',
    'cancelled',
  ].includes(value as string);
}

function isTerminalState(value: AgentControllerStateV1): boolean {
  return value === 'done' || value === 'failed' || value === 'cancelled';
}

function isControlAction(value: unknown): value is AgentTaskControlActionV1 {
  return value === 'cancel' || value === 'replan' || value === 'resume';
}

function isControlOutcome(value: unknown): value is AgentTaskControlOutcomeV1 {
  return value === 'cancelled' || value === 'replanRequired' || value === 'resumed';
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === 'object' && value !== null && !Array.isArray(value);
}

function hasExactKeys(value: Record<string, unknown>, expected: string[]): boolean {
  const keys = Object.keys(value).sort();
  const sortedExpected = [...expected].sort();
  return (
    keys.length === sortedExpected.length &&
    keys.every((key, index) => key === sortedExpected[index])
  );
}

File: crates/a3-protocol/src/agent_control.rs
use crate::{AgentControllerStateV1, ProtocolVersion};
use serde::{Deserialize, Serialize};
use std::fmt;

/// Requests current H11/E8 recovery controls for one selected durable task.
#[derive(Clone, Deserialize, Eq, PartialEq, Serialize)]
#[serde(deny_unknown_fields, rename_all = "camelCase")]
pub struct QueryAgentTaskRecoveryRequestV1 {
    protocol_version: ProtocolVersion,
    task_id: String,
}

impl fmt::Debug for QueryAgentTaskRecoveryRequestV1 {
    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
        formatter
            .debug_struct("QueryAgentTaskRecoveryRequestV1")
            .field("protocol_version", &self.protocol_version)
            .field("has_task_id", &!self.task_id.is_empty())
            .finish()
    }
}

impl QueryAgentTaskRecoveryRequestV1 {
    /// Returns the protocol version before the opaque task identity is parsed.
    #[must_use]
    pub const fn protocol_version(&self) -> ProtocolVersion {
        self.protocol_version
    }

    /// Returns the opaque task identity selected by the user.
    #[must_use]
    pub fn task_id(&self) -> &str {
        &self.task_id
    }
}

/// Closed explicit H11 recovery decision exposed to the untrusted WebView.
#[derive(Clone, Copy, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub enum AgentTaskControlActionV1 {
    /// Continue only after Core verifies current snapshots and evidence.
    Resume,
    /// Reopen stale work and require a new plan before further model work.
    Replan,
    /// Atomically terminate the selected current run.
    Cancel,
}

/// Applies one recovery choice against the exact Ledger anchors shown to the user.
#[derive(Clone, Deserialize, Eq, PartialEq, Serialize)]
#[serde(deny_unknown_fields, rename_all = "camelCase")]
pub struct ControlAgentTaskRunRequestV1 {
    protocol_version: ProtocolVersion,
    task_id: String,
    expected_ledger_revision: u32,
    expected_ledger_store_version: String,
    action: AgentTaskControlActionV1,
}

impl fmt::Debug for ControlAgentTaskRunRequestV1 {
    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
        formatter
            .debug_struct("ControlAgentTaskRunRequestV1")
            .field("protocol_version", &self.protocol_version)
            .field("has_task_id", &!self.task_id.is_empty())
            .field("expected_ledger_revision", &self.expected_ledger_revision)
            .field(
                "expected_ledger_store_version",
                &self.expected_ledger_store_version,
            )
            .field("action", &self.action)
            .finish()
    }
}

impl ControlAgentTaskRunRequestV1 {
    /// Returns the protocol version before any other request field is interpreted.
    #[must_use]
    pub const fn protocol_version(&self) -> ProtocolVersion {
        self.protocol_version
    }

    /// Returns the opaque task identity selected by the user.
    #[must_use]
    pub fn task_id(&self) -> &str {
        &self.task_id
    }

    /// Returns the exact visible Task Ledger revision used as an optimistic anchor.
    #[must_use]
    pub const fn expected_ledger_revision(&self) -> u32 {
        self.expected_ledger_revision
    }

    /// Returns the decimal optimistic Task Ledger store version.
    #[must_use]
    pub fn expected_ledger_store_version(&self) -> &str {
        &self.expected_ledger_store_version
    }

    /// Returns the explicit closed recovery decision.
    #[must_use]
    pub const fn action(&self) -> AgentTaskControlActionV1 {
        self.action
    }
}

/// Content-free current recovery state and exact optimistic control anchors.
#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(deny_unknown_fields, rename_all = "camelCase")]
pub struct AgentTaskRecoveryV1 {
    ledger_revision: u32,
    ledger_store_version: String,
    state: AgentControllerStateV1,
    run_snapshot_id: String,
    published_snapshot_id: String,
    snapshot_changed: bool,
    interrupted_tool_attempts: u32,
    stale_evidence_count: u32,
    mutation_reconciliation_required: bool,
    mutation_replan_required: bool,
    can_resume: bool,
}

impl AgentTaskRecoveryV1 {
    /// Creates one projection from already validated Core-owned recovery facts.
    #[allow(clippy::too_many_arguments)]
    #[must_use]
    pub fn new(
        ledger_revision: u32,
        ledger_store_version: String,
        state: AgentControllerStateV1,
        run_snapshot_id: String,
        published_snapshot_id: String,
        snapshot_changed: bool,
        interrupted_tool_attempts: u32,
        stale_evidence_count: u32,
        mutation_reconciliation_required: bool,
        mutation_replan_required: bool,
        can_resume: bool,
    ) -> Self {
        Self {
            ledger_revision,
            ledger_store_version,
            state,
            run_snapshot_id,
            published_snapshot_id,
            snapshot_changed,
            interrupted_tool_attempts,
            stale_evidence_count,
            mutation_reconciliation_required,
            mutation_replan_required,
            can_resume,
        }
    }
}

/// Expected bounded result of recovery inspection.
#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(deny_unknown_fields, rename_all = "camelCase", tag = "status")]
pub enum AgentTaskRecoveryResultV1 {
    /// No project is currently active in the Core.
    NoProject,
    /// The selected task no longer exists.
    TaskNotFound,
    /// The task exists but has no Task Ledger.
    LedgerUnavailable,
    /// Goal and Ledger revisions do not agree.
    GoalRevisionMismatch {
        /// Current immutable Goal Contract revision.
        current_revision: u32,
        /// Revision still materialized by the Ledger.
        ledger_revision: u32,
    },
    /// Durable state changed during the bounded read.
    ActivityChanged,
    /// No retained attempt has created a run.
    RunUnavailable,
    /// The latest run is terminal or no longer belongs to an active attempt.
    RunNotControllable {
        /// Last materialized finite controller state.
        state: AgentControllerStateV1,
    },
    /// Recovery controls are available against the returned exact anchors.
    Available {
        /// Current recovery projection.
        recovery: AgentTaskRecoveryV1,
    },
}

/// Strict envelope for current Agent task recovery controls.
#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(deny_unknown_fields, rename_all = "camelCase")]
pub struct AgentTaskRecoveryResponseV1 {
    protocol_version: ProtocolVersion,
    result: AgentTaskRecoveryResultV1,
}

impl AgentTaskRecoveryResponseV1 {
    /// Creates a response from an expected bounded Core result.
    #[must_use]
    pub const fn new(result: AgentTaskRecoveryResultV1) -> Self {
        Self {
            protocol_version: ProtocolVersion::CURRENT,
            result,
        }
    }

    /// Returns the explicit current recovery inspection state.
    #[must_use]
    pub const fn result(&self) -> &AgentTaskRecoveryResultV1 {
        &self.result
    }
}

/// Stable effect of one atomically committed recovery decision.
#[derive(Clone, Copy, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub enum AgentTaskControlOutcomeV1 {
    /// Recovery anchors were adopted and future continuation is allowed.
    Resumed,
    /// Invalid work was reopened and the normal replan path is now required.
    ReplanRequired,
    /// The run reached terminal Cancelled.
    Cancelled,
}

/// Expected bounded result of one explicit Agent task control command.
#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(deny_unknown_fields, rename_all = "camelCase", tag = "status")]
pub enum AgentTaskControlResultV1 {
    /// No project is currently active in the Core.
    NoProject,
    /// The selected task no longer exists.
    TaskNotFound,
    /// The task exists but has no Task Ledger.
    LedgerUnavailable,
    /// Goal and Ledger revisions do not agree.
    GoalRevisionMismatch {
        /// Current immutable Goal Contract revision.
        current_revision: u32,
        /// Revision still materialized by the Ledger.
        ledger_revision: u32,
    },
    /// Optimistic anchors changed before the command could commit.
    ActivityChanged,
    /// No retained attempt has created a run.
    RunUnavailable,
    /// The latest run is terminal or no longer belongs to an active attempt.
    RunNotControllable {
        /// Last materialized finite controller state.
        state: AgentControllerStateV1,
    },
    /// An Unknown mutation must first receive a full authoritative reconciliation.
    MutationReconciliationRequired,
    /// Resume is unsafe and requires the explicit Replan or Cancel choice.
    ResumeRequiresReplan,
    /// The requested choice was atomically committed.
    Applied {
        /// Stable effect of the committed choice.
        outcome: AgentTaskControlOutcomeV1,
        /// New optimistic Ledger store version.
        ledger_store_version: String,
        /// Resulting finite controller state.
        state: AgentControllerStateV1,
        /// Number of stale completed steps reopened.
        reopened_step_count: u32,
        /// Number of abandoned tool attempts marked interrupted.
        interrupted_tool_attempts: u32,
    },
}

/// Strict envelope for one explicit Agent task control command.
#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(deny_unknown_fields, rename_all = "camelCase")]
pub struct AgentTaskControlResponseV1 {
    protocol_version: ProtocolVersion,
    result: AgentTaskControlResultV1,
}

impl AgentTaskControlResponseV1 {
    /// Creates a response from an expected bounded Core result.
    #[must_use]
    pub const fn new(result: AgentTaskControlResultV1) -> Self {
        Self {
            protocol_version: ProtocolVersion::CURRENT,
            result,
        }
    }

    /// Returns the explicit result of the attempted recovery decision.
    #[must_use]
    pub const fn result(&self) -> &AgentTaskControlResultV1 {
        &self.result
    }
}

#[cfg(test)]
mod tests {
    use super::{AgentTaskControlActionV1, ControlAgentTaskRunRequestV1};
    use std::error::Error;

    #[test]
    fn control_request_is_closed_and_carries_no_run_or_snapshot_identity()
    -> Result<(), Box<dyn Error>> {
        let request: ControlAgentTaskRunRequestV1 = serde_json::from_value(serde_json::json!({
            "protocolVersion": 1,
            "taskId": "11".repeat(32),
            "expectedLedgerRevision": 2,
            "expectedLedgerStoreVersion": "7",
            "action": "cancel"
        }))?;
        assert_eq!(request.action(), AgentTaskControlActionV1::Cancel);
        let leaked = serde_json::from_value::<ControlAgentTaskRunRequestV1>(serde_json::json!({
            "protocolVersion": 1,
            "taskId": "11".repeat(32),
            "expectedLedgerRevision": 2,
            "expectedLedgerStoreVersion": "7",
            "action": "resume",
            "runId": "22".repeat(32)
        }));
        assert!(leaked.is_err());
        Ok(())
    }
}
